pilot/pkg/proxy/envoy/discovery.go

```diff
@@ -424,30 +436,61 @@ func debouncePush(startDebounce time.Time) {
 	if since > 2*DebounceAfter ||
 		time.Since(startDebounce) > DebounceMax {
 
-		log.Infof("Push debounce stable %d: %v since last change, %v since last push",
+		log.Infof("Push debounce stable %d: %v since last change, %v since last push, full=%v",
 			events,
-			since, time.Since(lastClearCache))
-		clearCacheMutex.Lock()
-		clearCacheTimerSet = false
-		lastClearCache = time.Now()
-		clearCacheMutex.Unlock()
-		V2ClearCache()
+			since, time.Since(lastClearCache), ds.fullPush)
+
+		ds.doPush()
+
 	} else {
-		log.Infof("Push debounce %d: %v since last change, %v since last push",
-			events,
-			since, time.Since(lastClearCache))
+		if debugDebounce {
+			log.Infof("Push debounce %d: %v since last change, %v since last push, full=%v",
+				events,
+				since, time.Since(lastClearCache), ds.fullPush)
+		}
 		time.AfterFunc(DebounceAfter, func() {
-			debouncePush(startDebounce)
+			ds.debouncePush(startDebounce)
 		})
 	}
 }
 
+// Start the actual push
+func (ds *DiscoveryService) doPush() {
+	// more config update events may happen while doPush is processing.
+	// we don't want to lose updates.
+	clearCacheMutex.Lock()
+
+	clearCacheTimerSet = false
+	lastClearCache = time.Now()
+	full := ds.fullPush
+	edsUpdates := BeforePush()
+
+	// Update the config values, next ConfigUpdate and eds updates will use this
+	ds.fullPush = false
+
+	clearCacheMutex.Unlock()
+
+	Push(full, edsUpdates)
+}
+
 // clearCache will clear all envoy caches. Called by service, instance and config handlers.
 // This will impact the performance, since envoy will need to recalculate.
 func (ds *DiscoveryService) clearCache() {
+	ds.ConfigUpdate(true)
+}
+
+// ConfigUpdate implements ConfigUpdater interface, used to request pushes.
+// It replaces the 'clear cache' from v1.
+func (ds *DiscoveryService) ConfigUpdate(full bool) {
 	clearCacheMutex.Lock()
 	defer clearCacheMutex.Unlock()
 
+	if full {
+		if !ds.fullPush && debugDebounce {
+			log.Info("Config change requiring full push")
```

Comment:
This should be debug. And when its debug, you can get rid of the debugDebounce completely

Reply:
It's just for quick troubleshooting. Setting log level to debug will generate too much garbage. I won't merge this to master - but while I still debug this it is quite useful.